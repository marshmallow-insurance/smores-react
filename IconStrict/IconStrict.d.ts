import { FC, FormEvent } from 'react';
import { Color } from '../theme';
import { MarginProps } from '../utils/space';
export type IconStrictProps = {
    /** className attribute to apply classes from props */
    className?: string;
    /** set size of the Icon (including background) */
    size?: 16 | 24 | 36 | 48;
    /** specify what Icon to render  */
    render: string;
    /** set icon colour */
    iconColor?: Color;
    /** set background colour */
    backgroundColor?: Color;
    /** function to handle click */
    handleClick?: (e: FormEvent<HTMLButtonElement>) => void;
    /** rotation degrees */
    rotate?: number;
} & MarginProps;
export declare const IconStrict: FC<IconStrictProps>;
