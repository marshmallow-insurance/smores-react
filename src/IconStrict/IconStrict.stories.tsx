import React from 'react'
import { IconStrict, IconStrictProps } from './IconStrict'

export default {
  title: 'IconStrict',
  component: IconStrict,
}

const Template = (props: IconStrictProps) => <IconStrict {...props} />

export const Default = Template.bind({})

Default.args = {
  size: 48,
  render: 'info',
  backgroundColor: 'mascarpone',
}

export const AlternateColours = Template.bind({})

AlternateColours.args = {
  size: 48,
  render: 'info',
  iconColor: 'cream',
  backgroundColor: 'marshmallowPink',
}

export const WithoutBackground = Template.bind({})

WithoutBackground.args = {
  size: 48,
  render: 'info',
}

export const WithClickHandler = Template.bind({})
WithClickHandler.args = {
  size: 48,
  render: 'info',
  handleClick: () => alert('clicked'),
  backgroundColor: 'mascarpone',
}
