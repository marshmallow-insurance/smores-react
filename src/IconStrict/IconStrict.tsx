import React, { ButtonHTMLAttributes, FC, FormEvent } from 'react'
import styled, { css } from 'styled-components'

import { darken } from 'polished'
import { Icon } from '../Icon'
import { Icons } from '../Icon/iconsList'

import { Color, theme } from '../theme'
import { focusOutlineStyle } from '../utils/focusOutline'
import { MarginProps } from '../utils/space'

export type IconStrictProps = {
  id?: string
  /** className attribute to apply classes from props */
  className?: string
  /** set size of the Icon (including background) */
  size?: 16 | 24 | 36 | 48
  /** specify what Icon to render  */
  render: Icons
  /** set icon colour */
  iconColor?: Color
  /** set background colour */
  backgroundColor?: Color
  /** function to handle click */
  handleClick?: (e: FormEvent<HTMLButtonElement>) => void
  /** rotation degrees */
  rotate?: number
} & MarginProps &
  Partial<ButtonHTMLAttributes<HTMLButtonElement>>

const iconSizes = {
  48: {
    smallSize: 28,
    padding: 10,
  },
  36: {
    smallSize: 20,
    padding: 8,
  },
  24: {
    smallSize: 12,
    padding: 6,
  },
  16: {
    smallSize: 10,
    padding: 3,
  },
}

export const IconStrict: FC<IconStrictProps> = ({
  id,
  className = '',
  size = 16,
  render,
  iconColor,
  backgroundColor,
  rotate,
  handleClick,
  title,
  ...otherProps
}) => {
  const isButton = !!handleClick
  const defaultLabel = isButton ? (title ? title : 'icon-button') : undefined
  return (
    <IconContainer
      id={id}
      as={isButton ? 'button' : 'div'}
      className={className}
      $size={size}
      $backgroundColor={backgroundColor}
      onClick={handleClick}
      title={defaultLabel}
      {...otherProps}
    >
      <Icon
        render={render}
        size={backgroundColor ? iconSizes[size].smallSize : size}
        color={iconColor}
        rotate={rotate}
      />
    </IconContainer>
  )
}

interface IIconStrict {
  $size: 16 | 24 | 36 | 48
  $backgroundColor?: Color
  onClick?: (e: FormEvent<HTMLButtonElement>) => void
}

const IconContainer = styled.div<IIconStrict>(
  ({ $size, $backgroundColor, onClick }) => css`
    position: relative;
    padding: ${$backgroundColor ? `${iconSizes[$size].padding}px` : 0};
    width: 100%;
    max-width: ${$size}px;
    height: ${$size}px;
    border-radius: 100%;
    background-color: ${$backgroundColor
      ? theme.colors[$backgroundColor]
      : 'none'};
    cursor: ${onClick ? 'pointer' : 'default'};

    ${onClick &&
    `
    &:hover {
      background-color: ${
        $backgroundColor ? darken(0.1, theme.colors[$backgroundColor]) : 'none'
      };
    }
      
    `}

    ${focusOutlineStyle}
  `,
)
