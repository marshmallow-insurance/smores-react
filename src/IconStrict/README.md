# < IconStrict />

## Props:

```ts
className?: string;
render:   string; /* icon set */
size?:    number; /* icon dimentions */
iconColor?:   string; /* from theme */
backgroundColor?: string; /* from theme */
rotate?:  number; /* degrees */
```

## Usage
This component is designed to provide strict sizes for the Icon component, and to render an optional background colour. If a background colour is provided, the Icon will be shrunk and the background will maintain the specified size.

```js
import {IconStrict} from '@mrshmllw/smores-react';

const App = () => (
  <IconStrict render='edit' backgroundColor="marshmallowPink" size={24} />
);
```
