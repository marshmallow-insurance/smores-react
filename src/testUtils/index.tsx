import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
