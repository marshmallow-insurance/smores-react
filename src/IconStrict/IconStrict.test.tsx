import React from 'react'
import { expect, it } from 'vitest'
import { render } from '../testUtils'
import { IconStrict } from './IconStrict'

describe('IconStrict', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<IconStrict render="car" />)
    expect(container).toMatchSnapshot()
  })
  it('renders correctly with background color', () => {
    const { container } = render(
      <IconStrict render="search" size={36} backgroundColor="matcha" />,
    )
    expect(container).toMatchSnapshot()
  })
})
