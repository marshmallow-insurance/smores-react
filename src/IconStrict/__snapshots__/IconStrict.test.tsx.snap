// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconStrict > renders correctly with background color 1`] = `
<div>
  <div
    class="sc-ggWZvA ibLTfl"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-fhHczv iNYmWQ"
      data-testid="search-container"
    >
      <svg
        fill="none"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 10a7 7 0 1114 0 7 7 0 01-14 0zm7-9a9 9 0 105.618 16.032l5.675 5.675a1 1 0 001.414-1.414l-5.675-5.675A9 9 0 0010 1z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`IconStrict > renders correctly with default props 1`] = `
<div>
  <div
    class="sc-ggWZvA iYvYSM"
  >
    <span
      class="sc-bRKDuR bHjpxP sc-fhHczv bkxtmk"
      data-testid="car-container"
    >
      <svg
        fill="none"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M21 10.424V5c0-2.757-2.243-5-5-5H8C5.243 0 3 2.243 3 5v5.424A5.003 5.003 0 000 15v4a1 1 0 001 1h1v3a1 1 0 002 0v-3h16v3a1 1 0 002 0v-3h1a1 1 0 001-1v-4a5.003 5.003 0 00-3-4.576zM5 13.995c-.552 0-1 .453-1 1.005a1 1 0 002 0v-.01a.996.996 0 00-1-.995zM8 2C6.346 2 5 3.346 5 5v5h14V5c0-1.654-1.346-3-3-3H8zm14 16H2v-3c0-1.654 1.346-3 3-3h14c1.654 0 3 1.346 3 3v3zm-4-3c0-.552.448-1.005 1-1.005s1 .443 1 .995V15a1 1 0 01-2 0z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
</div>
`;
